File: tests/load/k6/chat-api.js
import http from 'k6/http';
import { check, sleep } from 'k6';
import { Rate, Trend } from 'k6/metrics';

// Custom metrics
const errorRate = new Rate('errors');
const chatLatency = new Trend('chat_latency');

// Test configuration
export const options = {
  stages: [
    { duration: '30s', target: 10 },  // Ramp-up to 10 users
    { duration: '1m', target: 50 },   // Ramp-up to 50 users
    { duration: '2m', target: 50 },   // Stay at 50 users
    { duration: '30s', target: 100 }, // Spike to 100 users
    { duration: '1m', target: 100 },  // Stay at 100 users
    { duration: '30s', target: 0 },   // Ramp-down to 0 users
  ],
  thresholds: {
    http_req_duration: ['p(95)<500', 'p(99)<1000'], // 95% < 500ms, 99% < 1s
    http_req_failed: ['rate<0.01'],                  // Error rate < 1%
    errors: ['rate<0.05'],                           // Custom error rate < 5%
    chat_latency: ['p(95)<300'],                     // Chat API 95% < 300ms
  },
};

const BASE_URL = __ENV.API_URL || 'http://localhost:8100';

// Setup: Login and get token
export function setup() {
  const loginRes = http.post(`${BASE_URL}/auth/login`, JSON.stringify({
    username: 'admin',
    password: 'admin123',
  }), {
    headers: { 'Content-Type': 'application/json' },
  });

  const token = loginRes.json('token');
  return { token };
}

// Main test function
export default function (data) {
  const headers = {
    'Content-Type': 'application/json',
    'Authorization': `Bearer ${data.token}`,
  };

  // Test 1: Health check
  const healthRes = http.get(`${BASE_URL}/health`);
  check(healthRes, {
    'health check status is 200': (r) => r.status === 200,
    'health check has status field': (r) => r.json('status') !== undefined,
  });

  sleep(0.5);

  // Test 2: Chat API - Send message
  const chatPayload = JSON.stringify({
    content: `Test message ${__VU}-${__ITER}`,
  });

  const chatStart = Date.now();
  const chatRes = http.post(`${BASE_URL}/api/chat`, chatPayload, { headers });
  const chatDuration = Date.now() - chatStart;

  chatLatency.add(chatDuration);

  const chatSuccess = check(chatRes, {
    'chat status is 200': (r) => r.status === 200,
    'chat has message': (r) => r.json('message') !== undefined,
    'chat has conversation_id': (r) => r.json('conversation_id') !== undefined,
    'chat latency < 1s': (r) => r.timings.duration < 1000,
  });

  errorRate.add(!chatSuccess);

  sleep(1);

  // Test 3: Get conversations
  const conversationsRes = http.get(`${BASE_URL}/api/chat/conversations`, { headers });
  check(conversationsRes, {
    'conversations status is 200': (r) => r.status === 200,
    'conversations has array': (r) => Array.isArray(r.json('conversations')),
  });

  sleep(0.5);

  // Test 4: Memory search (if conversation exists)
  const searchRes = http.get(`${BASE_URL}/api/memory/search?q=test&limit=5`, { headers });
  check(searchRes, {
    'search status is 200': (r) => r.status === 200,
    'search has results': (r) => r.json('results') !== undefined,
  });

  sleep(1);
}

// Teardown
export function teardown(data) {
  console.log('Load test completed');
}

File: tests/load/locust/chat_api.py
"""
Locust load test for Jarvis Chat API
Run with: locust -f chat_api.py --host=http://localhost:8100
"""

from locust import HttpUser, task, between, events
import json
import time
from datetime import datetime

class JarvisUser(HttpUser):
    """
    Simulates a user interacting with Jarvis Chat API
    """
    wait_time = between(1, 3)  # Wait 1-3 seconds between tasks

    def on_start(self):
        """Login and get JWT token"""
        response = self.client.post("/auth/login", json={
            "username": "admin",
            "password": "admin123"
        })

        if response.status_code == 200:
            data = response.json()
            self.token = data.get("token")
            self.headers = {
                "Authorization": f"Bearer {self.token}",
                "Content-Type": "application/json"
            }
        else:
            self.token = None
            self.headers = {}

    @task(5)  # Weight: 5 (most common task)
    def send_chat_message(self):
        """Send a chat message"""
        if not self.token:
            return

        payload = {
            "content": f"Test message from user at {datetime.now().isoformat()}"
        }

        with self.client.post(
            "/api/chat",
            json=payload,
            headers=self.headers,
            catch_response=True
        ) as response:
            if response.status_code == 200:
                data = response.json()
                if "message" in data and "conversation_id" in data:
                    response.success()

                    # Record latency
                    latency = data.get("latency_ms", 0)
                    if latency > 1000:
                        response.failure(f"Latency too high: {latency}ms")
                else:
                    response.failure("Missing required fields in response")
            else:
                response.failure(f"Status code: {response.status_code}")

    @task(3)  # Weight: 3
    def list_conversations(self):
        """List all conversations"""
        if not self.token:
            return

        with self.client.get(
            "/api/chat/conversations",
            headers=self.headers,
            catch_response=True
        ) as response:
            if response.status_code == 200:
                data = response.json()
                if "conversations" in data:
                    response.success()
                else:
                    response.failure("Missing conversations field")
            else:
                response.failure(f"Status code: {response.status_code}")

    @task(2)  # Weight: 2
    def search_memory(self):
        """Search in memory"""
        if not self.token:
            return

        with self.client.get(
            "/api/memory/search?q=test&limit=10",
            headers=self.headers,
            catch_response=True
        ) as response:
            if response.status_code == 200:
                data = response.json()
                if "results" in data and "query_time_ms" in data:
                    query_time = data.get("query_time_ms", 0)
                    if query_time > 200:
                        response.failure(f"Search too slow: {query_time}ms")
                    else:
                        response.success()
                else:
                    response.failure("Missing required fields")
            else:
                response.failure(f"Status code: {response.status_code}")

    @task(1)  # Weight: 1
    def create_conversation(self):
        """Create a new conversation"""
        if not self.token:
            return

        payload = {
            "title": f"Test Conversation {time.time()}"
        }

        with self.client.post(
            "/api/chat/conversation",
            json=payload,
            headers=self.headers,
            catch_response=True
        ) as response:
            if response.status_code == 200:
                data = response.json()
                if "id" in data and "title" in data:
                    response.success()
                else:
                    response.failure("Missing required fields")
            else:
                response.failure(f"Status code: {response.status_code}")

    @task(1)  # Weight: 1
    def health_check(self):
        """Check API health"""
        with self.client.get(
            "/health",
            catch_response=True
        ) as response:
            if response.status_code == 200:
                data = response.json()
                if data.get("status") == "healthy":
                    response.success()
                else:
                    response.failure(f"Unhealthy status: {data.get('status')}")
            else:
                response.failure(f"Status code: {response.status_code}")


# Event listeners for custom metrics
@events.test_start.add_listener
def on_test_start(environment, **kwargs):
    print("=" * 60)
    print("Starting Jarvis Chat API Load Test")
    print(f"Target: {environment.host}")
    print("=" * 60)


@events.test_stop.add_listener
def on_test_stop(environment, **kwargs):
    print("=" * 60)
    print("Load test completed")
    print("=" * 60)
